### PR TITLE
Add `GenericListViewArray` for `ListView` & `LargeListView`

### PR DESCRIPTION
**Pull request by Kikkon**
# Which issue does this PR close?

A part of https://github.com/apache/arrow-rs/issues/5501

# Rationale for this change
 

# What changes are included in this PR?

Add GenericListViewArray implement

# Are there any user-facing changes?

### CODE REVIEW

**Comment by Kikkon** [comment]
@tustvold  If you have time, I would appreciate it if you could help me review this PR. Thanks!

File: arrow-array/src/array/list_view_array.rs [hunk 1]
```diff
@@ -0,0 +1,990 @@
...
+    /// Creates a [`GenericListViewArray`] from an iterator of primitive values
+    ///
+    pub fn from_iter_primitive<T, P, I>(_iter: I) -> Self
+    where
+        T: ArrowPrimitiveType,
+        P: IntoIterator<Item = Option<<T as ArrowPrimitiveType>::Native>>,
+        I: IntoIterator<Item = Option<P>>,
+    {
+        panic!("Not yet implemented, wait list view array builder to be implemented")
+    }
```
Comment (tustvold) [inline arrow-array/src/array/list_view_array.rs:309]: Perhaps we could just not provide this for now

File: arrow-array/src/array/list_view_array.rs [hunk 2]
```diff
@@ -0,0 +1,990 @@
...
+
+impl<OffsetSize: OffsetSizeTrait> From<FixedSizeListArray> for GenericListViewArray<OffsetSize> {
+    fn from(value: FixedSizeListArray) -> Self {
+        let (field, size) = match value.data_type() {
+            DataType::FixedSizeList(f, size) => (f, *size as usize),
+            _ => unreachable!(),
+        };
+        let iter = std::iter::repeat(size).take(value.len());
+        let mut offsets = Vec::with_capacity(iter.size_hint().0);
+        offsets.push(OffsetSize::usize_as(0));
```
Comment (tustvold) [inline arrow-array/src/array/list_view_array.rs:424]: Is this correct?

File: arrow-array/src/array/list_view_array.rs [hunk 3]
```diff
@@ -0,0 +1,990 @@
...
+        assert_eq!(6, list_array.value_offsets()[2]);
+        assert_eq!(2, list_array.value_sizes()[2]);
+        assert_eq!(2, list_array.value_length(2));
+        assert_eq!(
+            0,
+            list_array
+                .value(0)
+                .as_any()
+                .downcast_ref::<Int32Array>()
+                .unwrap()
```
Comment (tustvold) [inline arrow-array/src/array/list_view_array.rs:596]:
```suggestion
                .as_primitive::<Int32Type>()
```
And the same elsewhere

File: arrow-array/src/array/list_view_array.rs [hunk 4]
```diff
@@ -0,0 +1,990 @@
...
+        };
+        let iter = std::iter::repeat(size).take(value.len());
+        let mut offsets = Vec::with_capacity(iter.size_hint().0);
+        offsets.push(OffsetSize::usize_as(0));
+        let mut acc = 0_usize;
+        let iter = std::iter::repeat(size).take(value.len());
+        let mut sizes = Vec::with_capacity(iter.size_hint().0);
+        for size in iter {
+            acc = acc.checked_add(size).expect("usize overflow");
+            offsets.push(OffsetSize::usize_as(acc));
```
Comment (tustvold) [inline arrow-array/src/array/list_view_array.rs:430]:
```suggestion
            offsets.push(OffsetSize::usize_as(acc));
            acc = acc.checked_add(size).expect("usize overflow");
```

File: arrow-array/src/array/list_view_array.rs [hunk 5]
```diff
@@ -0,0 +1,990 @@
...
+        builder.values().append_slice(&[0, 0, 0]);
+        builder.append(false);
+        builder.values().append_slice(&[4, 5, 6]);
+        builder.append(true);
+        let list: ListViewArray = builder.finish().into();
+        let values: Vec<_> = list
+            .iter()
+            .map(|x| x.map(|x| x.as_primitive::<Int32Type>().values().to_vec()))
+            .collect();
+        assert_eq!(values, vec![Some(vec![1, 2, 3]), None, Some(vec![4, 5, 6])])
```
Comment (tustvold) [inline arrow-array/src/array/list_view_array.rs:968]: Could we add tests of the offsets and sizes buffers, I think they're currently not quite correct

**Review by tustvold — COMMENTED** [review]
Sorry for the delay, been swamped and then sick, but this is looking almost there

**Comment by Kikkon** [comment]
Thank you for taking the time to review my code despite being busy and unwell.

**Comment by tustvold** [comment]
For future reference, it is helpful for reviewers if you don't force push branches once reviews have been made, this makes it much easier to see what has changed

**Comment by Kikkon** [comment]
> For future reference, it is helpful for reviewers if you don't force push branches once reviews have been made, this makes it much easier to see what has changed. I'll try to find time to review this later today, but I now have to do a full review again which is quite time consuming on such a large PR

@tustvold  Thank you for your help. I will be more careful about the force push issue in my future commits. I am very sorry.

**Comment by Kikkon** [comment]
Hi @tustvold , perhaps when you're free this week, you could help me review this. This way, we can continue to move forward with this work. Thank you.

**Comment by tustvold** [comment]
I'm afraid I don't have capacity to assist with this initiative further, but perhaps @alamb may do

**Comment by Kikkon** [comment]
@alamb  Perhaps you can help advance this PR when they have time. 🥸

File: arrow-array/src/array/list_view_array.rs [hunk 6]
```diff
@@ -0,0 +1,938 @@
...
+            DataType::FixedSizeList(f, size) => (f, *size as usize),
+            _ => unreachable!(),
+        };
+        let mut offsets = Vec::with_capacity(size);
+        let mut acc = 0_usize;
+        let iter = std::iter::repeat(size).take(value.len());
+        let mut sizes = Vec::with_capacity(iter.size_hint().0);
+        for size in iter {
+            offsets.push(OffsetSize::usize_as(acc));
+            acc = acc.checked_add(size).expect("usize overflow");
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:414]: Minor corner case nit: Imagine we are working with `i32` offsets and we have `1Gi - 1` fixed size lists of size 2.  Will this panic even thought it shouldn't.

File: arrow-array/src/array/list_view_array.rs [hunk 7]
```diff
@@ -0,0 +1,938 @@
...
+            sizes.push(OffsetSize::usize_as(size));
+        }
+        OffsetSize::from_usize(acc).expect("offset overflow");
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:417]: Why this final check?  It seems redundant.

File: arrow-array/src/array/list_view_array.rs [hunk 8]
```diff
@@ -0,0 +1,938 @@
...
+            value_sizes: sizes,
+        }
+    }
+}
+
+impl<OffsetSize: OffsetSizeTrait> GenericListViewArray<OffsetSize> {
+    fn try_new_from_array_data(data: ArrayData) -> Result<Self, ArrowError> {
+        if data.buffers().len() != 2 {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "ListViewArray data should contain two buffer (value offsets & value size), had {}",
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:434]:
```suggestion
                "ListViewArray data should contain two buffers (value offsets & value sizes), had {}",
```

File: arrow-array/src/array/list_view_array.rs [hunk 9]
```diff
@@ -0,0 +1,938 @@
...
+        } else {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "[Large]ListViewArray's datatype must be [Large]ListViewArray(). It is {:?}",
+                data.data_type()
+            )));
+        }
+
+        let values = make_array(values);
+        // SAFETY:
+        // ArrayData is valid, and verified type above
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:466]: Why `SAFETY`?  We aren't doing anything unsafe here that I can see.

File: arrow-array/src/array/list_view_array.rs [hunk 10]
```diff
@@ -0,0 +1,938 @@
...
+            .iter()
+            .map(|x| x.map(|x| x.as_primitive::<Int32Type>().values().to_vec()))
+            .collect();
+        assert_eq!(values, vec![Some(vec![1, 2, 3]), None, Some(vec![4, 5, 6])]);
+        let offsets = list.value_offsets();
+        assert_eq!(offsets, &[0, 3, 6]);
+        let sizes = list.value_sizes();
+        assert_eq!(sizes, &[3, 3, 3]);
+    }
+}
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:938]: Some other tests might be:

 * Test where the lists overlap (e.g. sizes = `[5, 5]` and offsets = `[0, 3]` and `values` has length `8`.
 * Test where offsets does not cover the entire range of values (e.g. values has 50 things and offsets only go up to 10)
 * List of empty lists (with an empty values)

**Review by westonpace — COMMENTED** [review]
Some suggestions but overall this matches my understanding of list view arrays.  Thanks for the work!

File: arrow-array/src/array/list_view_array.rs [hunk 11]
```diff
@@ -0,0 +1,938 @@
...
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "Invalid offset and size values for {}ListViewArray, offset: {}, size: {}, values.len(): {}",
+                    OffsetSize::PREFIX, offset, size, values.len()
+                )));
+            }
+        }
+
+        if !field.is_nullable() && values.is_nullable() {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Non-nullable field of {}ListViewArray {:?} cannot contain nulls",
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs:133]: （： Aha, I directly referenced the error messages from the list array for this part to maintain consistency.

File: arrow-array/src/array/list_view_array.rs [hunk 12]
```diff
@@ -0,0 +1,938 @@
...
+
+    /// Create a new [`GenericListViewArray`] from the provided parts
+    ///
+    /// # Errors
+    ///
+    /// Errors if
+    ///
+    /// * `offsets.len() != sizes.len()`
+    /// * `offsets.len() != nulls.len()`
+    /// * `offsets.last() > values.len()`
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs:82]: I reviewed the code again and realized that I had already validated all the offsets, so I changed it to offsets[i] > values.len().

File: arrow-array/src/array/list_view_array.rs [hunk 13]
```diff
@@ -0,0 +1,938 @@
...
+    /// allowing for zero-copy cloning
+    #[inline]
+    pub fn sizes(&self) -> &ScalarBuffer<OffsetSize> {
+        &self.value_sizes
+    }
+
+    /// Returns a clone of the value type of this list.
+    pub fn value_type(&self) -> DataType {
+        self.values.data_type().clone()
+    }
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs:235]: This part is mainly to maintain consistency with the methods for other types in arrow-rs.

File: arrow-array/src/array/list_view_array.rs [hunk 14]
```diff
@@ -0,0 +1,938 @@
...
+
+    /// Returns the sizes values in the offsets buffer
+    #[inline]
+    pub fn value_sizes(&self) -> &[OffsetSize] {
+        &self.value_sizes
+    }
+
+    /// Returns the length for value at index `i`.
+    #[inline]
+    pub fn value_length(&self, i: usize) -> OffsetSize {
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs:267]: You are right! I changed the naming and completed the corresponding methods.

File: arrow-array/src/array/list_view_array.rs [hunk 15]
```diff
@@ -0,0 +1,938 @@
...
+        Arc::new(self.slice(offset, length))
+    }
+
+    fn len(&self) -> usize {
+        self.sizes().len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs:337]: I think we should use value size to verify emptiness here.

**Comment by Kikkon** [comment]
hi @westonpace please help me review again

File: arrow-array/src/array/list_view_array.rs [hunk 16]
```diff
@@ -0,0 +1,1025 @@
...
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Length of offsets buffer and sizes buffer must be equal for {}ListViewArray, got {} and {}",
+                OffsetSize::PREFIX, len, sizes.len()
+            )));
+        }
+
+        for (offset, size) in offsets.iter().zip(sizes.iter()) {
+            let offset = offset.as_usize();
+            let size = size.as_usize();
+            if offset.saturating_add(size) > values.len() {
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:102]: This should be a `checked_add`.

File: arrow-array/src/array/list_view_array.rs [hunk 17]
```diff
@@ -0,0 +1,1025 @@
...
+        };
+        let mut acc = 0_usize;
+        let iter = std::iter::repeat(size).take(value.len());
+        let mut sizes = Vec::with_capacity(iter.size_hint().0);
+        let mut offsets = Vec::with_capacity(iter.size_hint().0);
+        for size in iter {
+            offsets.push(OffsetSize::usize_as(acc));
+            acc = acc.saturating_add(size);
+            sizes.push(OffsetSize::usize_as(size));
+        }
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:400]: I did like that you were checking for overflow here, my only concern was that it was in the wrong spot.  What if, before the for loop, you checked to ensure that `value.len()` is less than or equal to the max offset?  Then you can keep using `saturating_add` here.

**Review by westonpace — COMMENTED** [review]
A few last small things.

**Comment by alamb** [comment]
I am depressed about the large review backlog in this crate. We are looking for more help from the community reviewing PRs -- see https://github.com/apache/arrow-rs/issues/6418 for more

**Comment by Kikkon** [comment]
@westonpace  😭 Apologies for the delay in processing this pull request due to work reasons. I hope you can help review it when you have time.

File: arrow-array/src/array/list_view_array.rs (hunk 17 above)
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs]: I've reconsidered, and it seems that no additional validation is needed here. 🤔

File: arrow-array/src/array/list_view_array.rs [hunk 18]
```diff
@@ -393,9 +403,11 @@ impl<OffsetSize: OffsetSizeTrait> From<FixedSizeListArray> for GenericListViewAr
         let iter = std::iter::repeat(size).take(value.len());
         let mut sizes = Vec::with_capacity(iter.size_hint().0);
         let mut offsets = Vec::with_capacity(iter.size_hint().0);
+        //check if the size is usize
+
```
Comment (westonpace) [inline arrow-array/src/array/list_view_array.rs:407]: Not entirely sure I understand the meaning of this comment.

**Review by westonpace — APPROVED** [review]
One minor thing but I think this is good to go.  Thanks for addressing the earlier reviews :)

File: arrow-array/src/array/list_view_array.rs (hunk 18 above)
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs]: I think this comment is no longer needed, so I've removed it.

File: arrow-array/src/array/list_view_array.rs [hunk 19]
```diff
@@ -0,0 +1,1036 @@
...
+                    OffsetSize::PREFIX
+                ))
+            })? > values.len()
+            {
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "Offset + size for {}ListViewArray must be within the bounds of the child array, got offset: {}, size: {}, child array length: {}",
+                    OffsetSize::PREFIX,
+                    offset,
+                    size,
+                    values.len()
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:115]:
```suggestion
                    "Offset + size for {}ListViewArray must be within the bounds of the child array, got offset: {offset}, size: {size}, child array length: {}",
                    OffsetSize::PREFIX,
                    values.len()
```

File: arrow-array/src/array/list_view_array.rs [hunk 20]
```diff
@@ -0,0 +1,1036 @@
...
+                    "Incorrect length of null buffer for {}ListViewArray, expected {len} got {}",
+                    OffsetSize::PREFIX,
+                    n.len(),
+                )));
+            }
+        }
+        if len != sizes.len() {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Length of offsets buffer and sizes buffer must be equal for {}ListViewArray, got {} and {}",
+                OffsetSize::PREFIX, len, sizes.len()
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:96]:
```suggestion
                "Length of offsets buffer and sizes buffer must be equal for {}ListViewArray, got {len} and {}",
                OffsetSize::PREFIX, sizes.len()
```

File: arrow-array/src/array/list_view_array.rs [hunk 21]
```diff
@@ -0,0 +1,1036 @@
...
+    pub fn value_size(&self, i: usize) -> OffsetSize {
+        self.value_sizes[i]
+    }
+
+    /// Returns the offset for value as index `i`.
+    pub fn value_offset(&self, i: usize) -> OffsetSize {
+        self.value_offsets[i]
+    }
+
+    /// constructs a new iterator
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:265]:
```suggestion
    /// Constructs a new iterator
```

File: arrow-array/src/array/list_view_array.rs [hunk 22]
```diff
@@ -0,0 +1,1036 @@
...
+    pub fn iter<'a>(&'a self) -> GenericListViewArrayIter<'a, OffsetSize> {
+        GenericListViewArrayIter::<'a, OffsetSize>::new(self)
+    }
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:268]: Can't we use lifetime elision?
```suggestion
    pub fn iter(&self) -> GenericListViewArrayIter<'_, OffsetSize> {
        GenericListViewArrayIter::<'_, OffsetSize>::new(self)
    }
```

File: arrow-array/src/array/list_view_array.rs [hunk 23]
```diff
@@ -0,0 +1,1036 @@
...
+
+        let values = data.child_data()[0].clone();
+
+        if let Some(child_data_type) = Self::get_type(data.data_type()) {
+            if values.data_type() != child_data_type {
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "[Large]ListViewArray's child datatype {:?} does not \
+                             correspond to the List's datatype {:?}",
+                    values.data_type(),
+                    child_data_type
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:448]:
```suggestion
                    "{}ListViewArray's child datatype {:?} does not \
                             correspond to the List's datatype {:?}",
                    OffsetSize::PREFIX,
                    values.data_type(),
                    child_data_type
```

File: arrow-array/src/array/list_view_array.rs [hunk 24]
```diff
@@ -0,0 +1,1036 @@
...
+                )));
+            }
+        } else {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "[Large]ListViewArray's datatype must be [Large]ListViewArray(). It is {:?}",
+                data.data_type()
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:454]:
```suggestion
                "{}ListViewArray's datatype must be {}ListViewArray(). It is {:?}",
                OffsetSize::PREFIX,
                OffsetSize::PREFIX,
                data.data_type()
```

File: arrow-array/src/array/list_view_array.rs [hunk 25]
```diff
@@ -0,0 +1,1036 @@
...
+use std::ops::Add;
+use std::sync::Arc;
+
+use crate::array::{make_array, print_long_array};
+use crate::iterator::GenericListViewArrayIter;
+use crate::{new_empty_array, Array, ArrayAccessor, ArrayRef, FixedSizeListArray, OffsetSizeTrait};
+
+/// A [`GenericListViewArray`] of variable size lists, storing offsets as `i32`.
+///
+// See [`ListViewBuilder`](crate::builder::ListViewBuilder) for how to construct a [`ListViewArray`]
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:31]:
```suggestion
/// See [`ListViewBuilder`](crate::builder::ListViewBuilder) for how to construct a [`ListViewArray`]
```

File: arrow-array/src/array/list_view_array.rs [hunk 26]
```diff
@@ -0,0 +1,1036 @@
...
+pub type ListViewArray = GenericListViewArray<i32>;
+
+/// A [`GenericListViewArray`] of variable size lists, storing offsets as `i64`.
+///
+// See [`LargeListViewBuilder`](crate::builder::LargeListViewBuilder) for how to construct a [`LargeListViewArray`]
```
Comment (mbrobbel) [inline arrow-array/src/array/list_view_array.rs:36]:
```suggestion
/// See [`LargeListViewBuilder`](crate::builder::LargeListViewBuilder) for how to construct a [`LargeListViewArray`]
```

File: arrow-array/src/array/list_view_array.rs [hunk 27]
```diff
@@ -0,0 +1,1036 @@
...
+    /// Returns ith value of this list view array.
+    /// # Safety
+    /// Caller must ensure that the index is within the array bounds
+    pub unsafe fn value_unchecked(&self, i: usize) -> ArrayRef {
+        let offset = self.value_offsets().get_unchecked(i).as_usize();
+        let length = self.value_sizes().get_unchecked(i).as_usize();
+        self.values.slice(offset, length)
+    }
+
+    /// Returns ith value of this list view array.
```
[the remark this answers is not in the document]
Comment (Kikkon) [reply arrow-array/src/array/list_view_array.rs:229]: @mbrobbel  Thank you for your review.

**Comment by alamb** [comment]
I took a quick skim of this PR and the code, tests and documentation are looking quite nice I think -- thank you @Kikkon  @westonpace and @mbrobbel 

🚀

**Pull request merged**